ui/pages/confirmations/hooks/alerts/transactions/useMultipleApprovalsAlerts.ts

```diff
@@ -0,0 +1,164 @@
+import { TransactionMeta } from '@metamask/transaction-controller';
+import { Hex } from '@metamask/utils';
+import { BigNumber } from 'bignumber.js';
+import { useMemo } from 'react';
+import { parseApprovalTransactionData } from '../../../../../../shared/modules/transaction.utils';
+import { RowAlertKey } from '../../../../../components/app/confirm/info/row/constants';
+import { Alert } from '../../../../../ducks/confirm-alerts/confirm-alerts';
+import { Severity } from '../../../../../helpers/constants/design-system';
+import { useI18nContext } from '../../../../../hooks/useI18nContext';
+import { useConfirmContext } from '../../../context/confirm';
+
+type ApprovalInfo = {
+  tokenAddress: Hex;
+  spenderAddress?: Hex;
+  amount?: BigNumber;
+  isNFT: boolean;
+};
+
+type TokenOutflow = {
+  tokenAddress: Hex;
+  amount: BigNumber;
+  isDecrease: boolean;
+};
+
+export function useMultipleApprovalsAlerts(): Alert[] {
+  const t = useI18nContext();
+  const { currentConfirmation } = useConfirmContext<TransactionMeta>();
+
+  const isBatchTransaction = currentConfirmation?.type === 'batch';
+  const nestedTransactions = currentConfirmation?.nestedTransactions;
+  const simulationDataArray =
+    currentConfirmation?.simulationData?.tokenBalanceChanges;
+
+  // Extract all approvals from nested transactions
+  const approvals = useMemo((): ApprovalInfo[] => {
+    if (!isBatchTransaction || !nestedTransactions) {
+      return [];
+    }
+
+    const approvalsList: ApprovalInfo[] = [];
+
+    nestedTransactions.forEach((transaction) => {
+      const { data, to } = transaction;
+      if (!data || !to) {
+        return;
+      }
+
+      const parseResult = parseApprovalTransactionData(data);
+      if (!parseResult) {
+        return;
+      }
+
+      const { name, amountOrTokenId, tokenAddress } = parseResult;
+
+      // Skip revocations (zero amounts or setApprovalForAll with false)
+      if (amountOrTokenId?.isZero() || parseResult.isRevokeAll) {
+        return;
+      }
+
+      let actualTokenAddress: Hex;
+      let spenderAddress: Hex | undefined;
+
+      switch (name) {
+        case 'approve':
+          if (tokenAddress) {
+            // Permit2 approve
+            actualTokenAddress = tokenAddress;
+            spenderAddress = to;
+          } else {
+            // Regular ERC20 approve
+            actualTokenAddress = to;
+          }
+          break;
+        case 'increaseAllowance':
+        case 'setApprovalForAll':
+          actualTokenAddress = to;
+          break;
+        default:
+          return;
+      }
+
+      const isNFT = name === 'setApprovalForAll';
+
+      approvalsList.push({
+        tokenAddress: actualTokenAddress,
+        spenderAddress,
+        amount: amountOrTokenId,
+        isNFT,
+      });
+    });
+
+    return approvalsList;
+  }, [isBatchTransaction, nestedTransactions]);
+
+  const tokenOutflows = useMemo((): TokenOutflow[] => {
```

Comment:
Can we omit these `useMemo` as they have no performance cost and we're not returning the objects?

We could instead split the logic into some local utility functions such as `getTokensWithDecrease(simulationData)`?